[JSC] Simplify Repatch Call IC

Constellation opened the pull request:
#### a38ef78e135b3a164a8361109797d5fc0fb7a9b5
<pre>
[JSC] Simplify Repatch Call IC
<a href="https://bugs.webkit.org/show_bug.cgi?id=274360">https://bugs.webkit.org/show_bug.cgi?id=274360</a>
<a href="https://rdar.apple.com/128339225">rdar://128339225</a>

Reviewed by Keith Miller.

This patch changes slow path implementation from generated special code to Data Call IC&apos;s one.

* Source/JavaScriptCore/bytecode/CallLinkInfo.cpp:
(JSC::CallLinkInfo::clearCallee):
(JSC::CallLinkInfo::revertCallToStub):
(JSC::CallLinkInfo::reset):
(JSC::CallLinkInfo::setVirtualCall):
(JSC::CallLinkInfo::emitFastPathImpl):
(JSC::CallLinkInfo::emitDataICFastPath):
(JSC::CallLinkInfo::emitTailCallDataICFastPath):
(JSC::CallLinkInfo::setStub):
(JSC::CallLinkInfo::emitFastPath):
(JSC::CallLinkInfo::emitTailCallFastPath):
(JSC::OptimizingCallLinkInfo::emitFastPath):
(JSC::OptimizingCallLinkInfo::emitTailCallFastPath):
(JSC::OptimizingCallLinkInfo::initializeFromDFGUnlinkedCallLinkInfo):
(JSC::CallLinkInfo::doneLocationIfExists): Deleted.
(JSC::OptimizingCallLinkInfo::setSlowPathCallDestination): Deleted.
(JSC::CallLinkInfo::emitSlowPathImpl): Deleted.
(JSC::CallLinkInfo::emitDataICSlowPath): Deleted.
(JSC::CallLinkInfo::emitSlowPath): Deleted.
(JSC::CallLinkInfo::emitTailCallSlowPath): Deleted.
(JSC::OptimizingCallLinkInfo::emitSlowPath): Deleted.
(JSC::OptimizingCallLinkInfo::emitTailCallSlowPath): Deleted.
* Source/JavaScriptCore/bytecode/CallLinkInfo.h:
(JSC::CallLinkInfo::codeOrigin const):
(JSC::UnlinkedCallLinkInfo::setDoneLocation): Deleted.
* Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
(JSC::InlineCacheCompiler::generateImpl):
(JSC::InlineCacheCompiler::emitProxyObjectAccess):
* Source/JavaScriptCore/bytecode/Repatch.cpp:
(JSC::linkSlowFor):
(JSC::linkPolymorphicCall):
* Source/JavaScriptCore/dfg/DFGJITCompiler.cpp:
(JSC::DFG::JITCompiler::link):
* Source/JavaScriptCore/dfg/DFGJITCompiler.h:
(JSC::DFG::JITCompiler::addJSCall): Deleted.
(JSC::DFG::JITCompiler::JSCallRecord::JSCallRecord): Deleted.
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::emitCall):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::emitCall):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/jit/JITCall.cpp:
(JSC::JIT::compileTailCall):
(JSC::JIT::compileOpCall):
* Source/JavaScriptCore/jit/JITThunks.h:
* Source/JavaScriptCore/jit/PolymorphicCallStubRoutine.cpp:
(JSC::PolymorphicCallNode::unlinkOrUpgradeImpl):
* Source/JavaScriptCore/jit/ThunkGenerators.cpp:
(JSC::polymorphicRepatchThunk): Deleted.
* Source/JavaScriptCore/jit/ThunkGenerators.h:
* Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
(JSC::Wasm::wasmToJS):

Canonical link: <a href="https://commits.webkit.org/279391@main">https://commits.webkit.org/279391@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c837f21a67dba48ef994356d903bb8fc69572d52

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53221 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32569 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5710 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56500 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3947 "Built successfully") 
| | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/39798 "Hash c837f21a for PR 28764 does not build (failure)") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3684 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43145 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2565 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55319 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/49/builds/39798 "Hash c837f21a for PR 28764 does not build (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45951 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24276 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/49/builds/39798 "Hash c837f21a for PR 28764 does not build (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3277 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2103 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/46577 "Built successfully and passed tests") | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/49/builds/39798 "Hash c837f21a for PR 28764 does not build (failure)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3433 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58096 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/52734 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28366 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3402 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50545 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29582 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46172 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49864 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30503 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/65039 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7846 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29340 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12383 "Found 8 new JSC stress test failures: stress/proxy-set-prototype-of.js.mini-mode, wasm.yaml/wasm/function-tests/br-if-loop-less-than.js.wasm-eager, wasm.yaml/wasm/function-tests/exceptions.js.wasm-eager, wasm.yaml/wasm/function-tests/rotl.js.wasm-eager, wasm.yaml/wasm/js-api/global-mutate.js.wasm-eager, wasm.yaml/wasm/regress/242294.js.wasm-eager, wasm.yaml/wasm/stress/exception-thrown-out-of-wasm.js.wasm-eager, wasm.yaml/wasm/stress/immutable-globals.js.wasm-eager (failure)") | 
<!--EWS-Status-Bubble-End-->

Achierius commented on Source/JavaScriptCore/bytecode/CallLinkInfo.cpp:
```diff
@@ -266,16 +261,13 @@ void CallLinkInfo::revertCallToStub()
-        MacroAssembler::repatchPointer(u.codeIC.m_codeBlockLocation, nullptr);
-        CCallHelpers::revertJumpReplacementToBranchPtrWithPatch(CCallHelpers::startOfBranchPtrWithPatchOnRegister(u.codeIC.m_calleeLocation), BaselineJITRegisters::Call::calleeGPR, nullptr);
-#else
-        RELEASE_ASSERT_NOT_REACHED();
```
Why does this code (i.e. a non-data IC, but with the JIT disabled) become reachable with this patch?

Achierius replied:
Answer is that the data/non-data ICs are now one and the same

Achierius commented on Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
```diff
@@ -2855,19 +2855,14 @@ void InlineCacheCompiler::generateImpl(unsigned index, AccessCase& accessCase)
+#if USE(JSVALUE64)
+            if (!codeBlock->useDataIC())
+                useDataIC = CallLinkInfo::UseDataIC::No;
+#endif
```
nit: I feel like this would be clearer as

```
#if USE(JSVALUE64)
     CallLinkInfo::UseDataIC useDataIC = codeBlock->useDataIC() ? CallLinkInfo::UseDataIC::Yes : CallLinkInfo::UseDataIC::No;
#else
     CallLinkInfo::UseDataIC useDataIC = CallLinkInfo::UseDataIC::No;
#endif
```

kmiller68 replied:
Why are we only disabling dataIC for JSValue64? I'm confused by this code.

Constellation replied:
It seems that 32bit non-DataIC is not working correclty. But anyway, we can just keep it broken.

Achierius commented on Source/JavaScriptCore/bytecode/CallLinkInfo.cpp:
```diff
@@ -365,16 +356,11 @@ JSGlobalObject* CallLinkInfo::globalObjectForSlowPath(JSCell* owner)
-}
-
-std::tuple<MacroAssembler::JumpList, MacroAssembler::Label> CallLinkInfo::emitFastPathImpl(CallLinkInfo* callLinkInfo, CCallHelpers& jit, UseDataIC useDataIC, bool isTailCall, ScopedLambda<void()>&& prepareForTailCall)
+void CallLinkInfo::emitFastPathImpl(CallLinkInfo* callLinkInfo, CCallHelpers& jit, UseDataIC useDataIC, bool isTailCall, ScopedLambda<void()>&& prepareForTailCall)
```
Since we know whether a tail call is being emitted anytime we call this function, could it be split into two (e.g. `emitTailCallFastPathImpl`, `emitFastPathImpl`)? It would avoid the iffy `nullptr` passing in those cases where `isTailCall = false`.

Constellation replied:
`nullptr` is passed when it is DataOnly CallLinkInfo.

kmiller68 commented on Source/JavaScriptCore/bytecode/CallLinkInfo.cpp:
```diff
@@ -159,8 +159,6 @@ void CallLinkInfo::setMonomorphicCallee(VM& vm, JSCell* owner, JSObject* callee,
         MacroAssembler::repatchPointer(u.codeIC.m_codeBlockLocation, codeBlock);
         MacroAssembler::repatchPointer(u.codeIC.m_calleeLocation, callee);
-#else
-        RELEASE_ASSERT_NOT_REACHED();
```
Why are we removing these asserts? Wouldn't using `!isDataIC()` when the JIT is disabled be wrong?

Constellation replied:
Added.

kmiller68 commented on Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
```diff
@@ -3597,11 +3582,14 @@ void InlineCacheCompiler::emitProxyObjectAccess(unsigned index, ProxyObjectAcces
+#if USE(JSVALUE64)
+    if (!codeBlock->useDataIC())
+        useDataIC = CallLinkInfo::UseDataIC::No;
+#endif
```
Ditto.

Constellation replied:
Ditto.

kmiller68 reviewed (approved):
r=me.

Constellation committed:
[JSC] Simplify Repatch Call IC
https://bugs.webkit.org/show_bug.cgi?id=274360
rdar://128339225

Reviewed by Keith Miller.

This patch changes slow path implementation from generated special code to Data Call IC's one.

* Source/JavaScriptCore/bytecode/CallLinkInfo.cpp:
(JSC::CallLinkInfo::clearCallee):
(JSC::CallLinkInfo::revertCallToStub):
(JSC::CallLinkInfo::reset):
(JSC::CallLinkInfo::setVirtualCall):
(JSC::CallLinkInfo::emitFastPathImpl):
(JSC::CallLinkInfo::emitDataICFastPath):
(JSC::CallLinkInfo::emitTailCallDataICFastPath):
(JSC::CallLinkInfo::setStub):
(JSC::CallLinkInfo::emitFastPath):
(JSC::CallLinkInfo::emitTailCallFastPath):
(JSC::OptimizingCallLinkInfo::emitFastPath):
(JSC::OptimizingCallLinkInfo::emitTailCallFastPath):
(JSC::OptimizingCallLinkInfo::initializeFromDFGUnlinkedCallLinkInfo):
(JSC::CallLinkInfo::doneLocationIfExists): Deleted.
(JSC::OptimizingCallLinkInfo::setSlowPathCallDestination): Deleted.
(JSC::CallLinkInfo::emitSlowPathImpl): Deleted.
(JSC::CallLinkInfo::emitDataICSlowPath): Deleted.
(JSC::CallLinkInfo::emitSlowPath): Deleted.
(JSC::CallLinkInfo::emitTailCallSlowPath): Deleted.
(JSC::OptimizingCallLinkInfo::emitSlowPath): Deleted.
(JSC::OptimizingCallLinkInfo::emitTailCallSlowPath): Deleted.
* Source/JavaScriptCore/bytecode/CallLinkInfo.h:
(JSC::CallLinkInfo::codeOrigin const):
(JSC::UnlinkedCallLinkInfo::setDoneLocation): Deleted.
* Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
(JSC::InlineCacheCompiler::generateImpl):
(JSC::InlineCacheCompiler::emitProxyObjectAccess):
* Source/JavaScriptCore/bytecode/Repatch.cpp:
(JSC::linkSlowFor):
(JSC::linkPolymorphicCall):
* Source/JavaScriptCore/dfg/DFGJITCompiler.cpp:
(JSC::DFG::JITCompiler::link):
* Source/JavaScriptCore/dfg/DFGJITCompiler.h:
(JSC::DFG::JITCompiler::addJSCall): Deleted.
(JSC::DFG::JITCompiler::JSCallRecord::JSCallRecord): Deleted.
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::emitCall):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::emitCall):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/jit/JITCall.cpp:
(JSC::JIT::compileTailCall):
(JSC::JIT::compileOpCall):
* Source/JavaScriptCore/jit/JITThunks.h:
* Source/JavaScriptCore/jit/PolymorphicCallStubRoutine.cpp:
(JSC::PolymorphicCallNode::unlinkOrUpgradeImpl):
* Source/JavaScriptCore/jit/ThunkGenerators.cpp:
(JSC::polymorphicRepatchThunk): Deleted.
* Source/JavaScriptCore/jit/ThunkGenerators.h:
* Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
(JSC::Wasm::wasmToJS):

Canonical link: https://commits.webkit.org/279391@main
